Helm Release Support

viveklak opened the pull request:
Fixes https://github.com/pulumi/home/issues/1811

lblackstone commented on themes/default/content/blog/helm-release-resource-for-kubernetes-generally-available/index.md:
```diff
@@ -0,0 +1,415 @@
+If you aren't already familiar with using Pulumi with Kubernetes, head on over to [Pulumi's getting started guide for Kubernetes]({{< relref "/docs/get-started/kubernetes/" >}}) first.
+
+1. In your chosen Kubernetes Pulumi project, make sure the referenced version of the Pulumi-Kubernetes SDK is at least `v3.15.0`.
+2. Helm Release support is configured to target the relevant cluster pointed to by the Kubernetes provider associated with it and no additional configuration is required by default. However, more advanced configuration options are supported, see the provider configuration [documentation]({{< relref "/registry/packages/kubernetes/api-docs/provider/#helmdriver_nodejs" >}}) for more details.
```
```suggestion
2. Like other Kubernetes resources managed by Pulumi, the target cluster for Helm Release is configured with a [Provider]({{< relref "/registry/packages/kubernetes/api-docs/provider">}}). The default options work for typical cases, and we also provide [advanced configuration options]({{< relref "/registry/packages/kubernetes/api-docs/provider/#helmdriver_nodejs" >}}) for private registries, etc.
```

lblackstone commented on themes/default/content/blog/helm-release-resource-for-kubernetes-generally-available/index.md:
```diff
@@ -0,0 +1,415 @@
+2. Helm Release support is configured to target the relevant cluster pointed to by the Kubernetes provider associated with it and no additional configuration is required by default. However, more advanced configuration options are supported, see the provider configuration [documentation]({{< relref "/registry/packages/kubernetes/api-docs/provider/#helmdriver_nodejs" >}}) for more details.
+3. Refer to the instructions for your chosen language in the [resource documentation]({{< relref "/registry/packages/kubernetes/api-docs/helm/v3/release/#create" >}}) to create a Helm Release resource.
+
+Lets look at a concrete example of Helm Release in action. In the following snippet we install [Redis](https://redis.io/) through a Helm Chart in each of the supported Pulumi languages using the new Helm Release resource:
```
```suggestion
Let's look at a concrete example of Helm Release in action. In the following snippet we install [Redis](https://redis.io/) using the new Helm Release resource:
```

lblackstone commented on themes/default/content/blog/helm-release-resource-for-kubernetes-generally-available/index.md:
```diff
@@ -0,0 +1,415 @@
+});
+
+
+// We can lookup resources once the release is installed. The release's
```
```suggestion
// We can look up resources once the release is installed. The release's
```

lblackstone commented on themes/default/content/blog/helm-release-resource-for-kubernetes-generally-available/index.md:
```diff
@@ -0,0 +1,415 @@
+				},
+			},
+
+			// By default Release resource will wait till all created resources
```
```suggestion
			// By default Release resource will wait until all created resources
```

lblackstone commented on themes/default/content/blog/helm-release-resource-for-kubernetes-generally-available/index.md:
```diff
@@ -0,0 +1,415 @@
+            create: true,
+        }
+    },
+    // By default Release resource will wait till all created resources
```
```suggestion
    // By default Release resource will wait until all created resources
```

lblackstone commented on themes/default/content/blog/helm-release-resource-for-kubernetes-generally-available/index.md:
```diff
@@ -0,0 +1,415 @@
+			return err
+		}
+
+		// We can lookup resources once the release is installed. The release's
```
```suggestion
		// We can look up resources once the release is installed. The release's
```

lblackstone commented on themes/default/content/blog/helm-release-resource-for-kubernetes-generally-available/index.md:
```diff
@@ -0,0 +1,415 @@
+		// We can lookup resources once the release is installed. The release's
+		// status field is set once the installation completes, so this, combined
+		// with `skipAwait: pulumi.BoolPtr(false)` above, will wait to retrieve
+        //  the Redis master ClusterIP till all resources in the Chart are available.
```
```suggestion
        //  the Redis master ClusterIP until all resources in the Chart are available.
```

lblackstone commented on themes/default/content/blog/helm-release-resource-for-kubernetes-generally-available/index.md:
```diff
@@ -0,0 +1,415 @@
+            "create": True,
+        },
+    },
+    # By default Release resource will wait till all created resources
```
```suggestion
    # By default Release resource will wait until all created resources
```

lblackstone commented on themes/default/content/blog/helm-release-resource-for-kubernetes-generally-available/index.md:
```diff
@@ -0,0 +1,415 @@
+
+release = Release("redis-helm", args=release_args)
+
+# We can lookup resources once the release is installed. The release's
```
```suggestion
# We can look up resources once the release is installed. The release's
```

lblackstone commented on themes/default/content/blog/helm-release-resource-for-kubernetes-generally-available/index.md:
```diff
@@ -0,0 +1,415 @@
+# We can lookup resources once the release is installed. The release's
+# status field is set once the installation completes, so this, combined
+# with `skip_await=False` above, will wait to retrieve the Redis master
+# ClusterIP till all resources in the Chart are available.
```
```suggestion
# ClusterIP until all resources in the Chart are available.
```

lblackstone commented on themes/default/content/blog/helm-release-resource-for-kubernetes-generally-available/index.md:
```diff
@@ -0,0 +1,415 @@
+                    ["create"] = true,
+                }
+            },
+            // By default Release resource will wait till all created resources
```
```suggestion
            // By default Release resource will wait until all created resources
```

lblackstone commented on themes/default/content/blog/helm-release-resource-for-kubernetes-generally-available/index.md:
```diff
@@ -0,0 +1,415 @@
+
+        var release = new Release("redis-helm", releaseArgs);
+
+        // We can lookup resources once the release is installed. The release's
```
```suggestion
        // We can look up resources once the release is installed. The release's
```

lblackstone commented on themes/default/content/blog/helm-release-resource-for-kubernetes-generally-available/index.md:
```diff
@@ -0,0 +1,415 @@
+        // We can lookup resources once the release is installed. The release's
+        // status field is set once the installation completes, so this, combined
+        // with `skip_await=False` above, will wait to retrieve the Redis master
+        // ClusterIP till all resources in the Chart are available.
```
```suggestion
        // ClusterIP until all resources in the Chart are available.
```

lblackstone commented on themes/default/content/docs/guides/adopting/from_kubernetes.md:
```diff
@@ -356,32 +356,31 @@ Resources:
+
+### Emulating Helm Charts With Chart Resources
+
+With the [Helm V2]({{< relref "/registry/packages/kubernetes/api-docs/helm/v2/release" >}}) and [Helm V3]({{< relref "/registry/packages/kubernetes/api-docs/helm/v3/release" >}}) chart resources, Pulumi renders the templates and applies them directly, much like with `ConfigFile` and `ConfigGroup` shown earlier, which means all provisioning happens client-side using your Kubernetes authentication  ...
```
I wonder if we should drop mentions of Helm v2 here while we're at it

viveklak replied:
Done.

lblackstone commented on themes/default/content/docs/guides/adopting/from_kubernetes.md:
```diff
@@ -392,9 +391,11 @@ let k8s = require("@pulumi/kubernetes");
 
 // Deploy the latest version of the stable/wordpress chart.
```
Several comments still refer to `stable/wordpress`

viveklak replied:
Updated.

lblackstone commented on themes/default/content/docs/guides/adopting/from_kubernetes.md:
```diff
@@ -557,6 +565,202 @@ $ curl http://$(pulumi stack output frontendIp)
 
+### Natively installing Helm Charts as Releases
+
+A new [Helm Release]({{< relref "/registry/packages/kubernetes/api-docs/helm/v3/release">}})) resource (GA as of [v3.15.0](https://github.com/pulumi/pulumi-kubernetes/releases/tag/v3.15.0) of the Pulumi Kubernetes Provider and SDK) is available in addition to the `Chart` resources discussed earlier. In this case, the Pulumi Kubernetes provider uses an embedded version of the Helm SDK to natively  ...
```
```suggestion
The [Helm Release]({{< relref "/registry/packages/kubernetes/api-docs/helm/v3/release">}})) resource (GA as of [v3.15.0](https://github.com/pulumi/pulumi-kubernetes/releases/tag/v3.15.0) of the Pulumi Kubernetes Provider) is another option for installing Charts. In this case, the Pulumi Kubernetes provider uses an embedded version of the Helm SDK to provide full-fidelity support for managing [`Helm Releases`](https://helm.sh/docs/glossary/#release).
```

lblackstone commented on themes/default/content/docs/guides/adopting/from_kubernetes.md:
```diff
@@ -557,6 +565,202 @@ $ curl http://$(pulumi stack output frontendIp)
+
+The `Release` resource type's inputs closely mirror the options supported by the Helm CLI and deviate slightly from the API supported by the `Chart` resources. Some key options are highlighted here:
+
+* `chart`: The required chart name (for instance, `"wordpress"`). In case of a local helm chart, a path can be specified as well.
```
```suggestion
* `chart`: The required chart name (for instance, `"wordpress"`). For a local helm chart, a path can be specified instead.
```

lblackstone commented on themes/default/content/docs/guides/adopting/from_kubernetes.md:
```diff
@@ -557,6 +565,202 @@ $ curl http://$(pulumi stack output frontendIp)
+
+For more details on all the supported inputs, please refer to the [API reference documentation]({{< relref "/registry/packages/kubernetes/api-docs/helm/v3/release/#inputs" >}}).
+
+Unlike `Chart` resource types, `Release` doesn't include references to the underlying Kubernetes resources created during the installation. As a result, just the `Release` resource is encoded in Pulumi state by default. The `Release` resource type includes the [`ReleaseStatus`]({{< relref "/registry/packages/kubernetes/api-docs/helm/v3/release/#releasestatus" >}}) as an output type. The `namespac ...
```
```suggestion
Unlike `Chart` resource types, `Release` doesn't include references to the underlying Kubernetes resources created during the installation. As a result, only the `Release` resource is stored in Pulumi state by default. The `Release` resource type includes the [`ReleaseStatus`]({{< relref "/registry/packages/kubernetes/api-docs/helm/v3/release/#releasestatus" >}}) as an output type. The following example shows how to retrieve resources managed by a Release.
```

lblackstone commented on themes/default/content/docs/guides/adopting/from_kubernetes.md:
```diff
@@ -557,6 +565,202 @@ $ curl http://$(pulumi stack output frontendIp)
+
+#### Installing a Helm Release
+
+To illustrate provisioning a Helm Chart using Pulumi, we will deploy the same `wordpress` chart as we did using the Helm `Chart` using `Release` instead:
```
```suggestion
We will deploy the same `wordpress` chart as before:
```

lblackstone commented on themes/default/content/docs/guides/adopting/from_kubernetes.md:
```diff
@@ -557,6 +565,202 @@ $ curl http://$(pulumi stack output frontendIp)
+                Repo = "https://charts.bitnami.com/bitnami"
+            },
+            Version = "9.6.0",
+        });
```
This version doesn't include the ClusterIP config

viveklak replied:
Removed from the python version instead.

lblackstone commented on themes/default/content/docs/guides/adopting/from_kubernetes.md:
```diff
@@ -557,6 +565,202 @@ $ curl http://$(pulumi stack output frontendIp)
+        version="9.6.0",
+        values={
+            "service": {
+                "type": "ClusterIP",
```
This example doesn't match the others.

lblackstone commented on themes/default/content/docs/guides/adopting/from_kubernetes.md:
````diff
@@ -557,6 +565,202 @@ $ curl http://$(pulumi stack output frontendIp)
+wpdev-xxxxxxxx  default         1               2022-02-01 01:06:08.513501 -0800 PST    deployed        wordpress-....
+```
+
+In addition, unlike the Chart resources, existing Helm releases deployed using the Helm CLI can be imported readily using the `pulumi import` command.
````
```suggestion
The Helm Release resource also supports importing existing Helm releases by using the `pulumi import` command.
```

lblackstone commented on themes/default/content/registry/packages/kubernetes/how-to-guides/choosing-the-right-helm-resource-for-your-use-case.md:
```diff
@@ -0,0 +1,101 @@
+
+In September 2021 we announced the **public preview** of a new [`Helm Release`]({{< relref "/registry/packages/kubernetes/api-docs/helm/v3/release">}}) which adds an additional option to the mix for Pulumi's Kubernetes users. As of [v3.15.0](https://github.com/pulumi/pulumi-kubernetes/releases/tag/v3.15.0) of the Pulumi Kubernetes SDK and Provider, this resource is now **Generally Available**.
+
+This guide discusses in some detail each of these options and provides a framework to determine using the right resource for your use case.
```
```suggestion
This guide should help you to choose the best option for your use case.
```

lblackstone commented on themes/default/content/registry/packages/kubernetes/how-to-guides/choosing-the-right-helm-resource-for-your-use-case.md:
```diff
@@ -0,0 +1,101 @@
+
+## Helm Chart Resources
+
+The [`Helm V2 Chart`]({{< relref "/registry/packages/kubernetes/api-docs/helm/v2/chart">}}), and [`Helm V3 Chart`]({{< relref "/registry/packages/kubernetes/api-docs/helm/v3/chart">}}) resources render the templates from your chart and applies them directly, which means all provisioning happens client-side using your Kubernetes authentication setup without needing a server-side component such as  ...
```
```suggestion
The [`Helm V2 Chart`]({{< relref "/registry/packages/kubernetes/api-docs/helm/v2/chart">}}), and [`Helm V3 Chart`]({{< relref "/registry/packages/kubernetes/api-docs/helm/v3/chart">}}) resources render the templates from your chart and then manage them directly with the Pulumi Kubernetes provider. These resources are implemented as [`Component Resources`]({{< relref "/docs/intro/concepts/resources/components/" >}}) which provide a number of benefits for Pulumi users:
```

lblackstone commented on themes/default/content/registry/packages/kubernetes/how-to-guides/choosing-the-right-helm-resource-for-your-use-case.md:
```diff
@@ -0,0 +1,101 @@
+
+### Benefits
+
+1. Visibility into all resources encapsulated by the Chart in Pulumi's state, allowing users the ability to directly query properties of individual resources.
```
```suggestion
1. Visibility into all resources encapsulated by the Chart in Pulumi's state, allowing users to directly query properties of individual resources.
```

lblackstone commented on themes/default/content/registry/packages/kubernetes/how-to-guides/choosing-the-right-helm-resource-for-your-use-case.md:
```diff
@@ -0,0 +1,101 @@
+### Benefits
+
+1. Visibility into all resources encapsulated by the Chart in Pulumi's state, allowing users the ability to directly query properties of individual resources.
+2. The above also allows Pulumi's Policy-as-Code framework - [`CrossGuard`]({{< relref "/docs/guides/crossguard/" >}}) to enforce policies on all resources installed by Helm charts
```
```suggestion
2. Tight integration with Pulumi's Policy-as-Code framework - [`CrossGuard`]({{< relref "/docs/guides/crossguard/" >}}) to enforce policies on all resources installed by Helm charts
```

lblackstone commented on themes/default/content/registry/packages/kubernetes/how-to-guides/choosing-the-right-helm-resource-for-your-use-case.md:
```diff
@@ -0,0 +1,101 @@
+
+1. Visibility into all resources encapsulated by the Chart in Pulumi's state, allowing users the ability to directly query properties of individual resources.
+2. The above also allows Pulumi's Policy-as-Code framework - [`CrossGuard`]({{< relref "/docs/guides/crossguard/" >}}) to enforce policies on all resources installed by Helm charts
+3. Ability to leverage transformations to programmatically manipulate resources installed by Helm charts in any of the Pulumi supported programming languages
```
Would be nice to add a link to more info on transformations

viveklak replied:
Done.

lblackstone commented on themes/default/content/registry/packages/kubernetes/how-to-guides/choosing-the-right-helm-resource-for-your-use-case.md:
```diff
@@ -0,0 +1,101 @@
+2. Existing Helm releases installed via the Helm CLI can be imported into Pulumi state as of [v3.12.1](https://github.com/pulumi/pulumi-kubernetes/releases/tag/v3.12.1) of the Pulumi Kubernetes SDK
+3. Releases installed via Pulumi are serialized by the chosen Helm driver in the cluster and can be queried by the Helm CLI.
+
+However, in return for these improvements, Helm Release resource incurs a few inherrent limitations:
```
```suggestion
```

lblackstone commented on themes/default/content/registry/packages/kubernetes/how-to-guides/choosing-the-right-helm-resource-for-your-use-case.md:
```diff
@@ -0,0 +1,101 @@
+### Limitations
+
+1. The ability to support transformations is limited to that offered by the `Helm` CLI (running `helm install` with `--post-renderer postrenderer`)
+2. Pulumi doesn't have direct access to the underlying Kubernetes resources installed by `Helm`, as a result, `CrossGuard` policies can't be enforced on Helm Release resources.
```
```suggestion
2. Since Pulumi is not directly managing the underlying Kubernetes resource installed by `Helm`, `CrossGuard` policies can't be enforced on these resources.
```

lblackstone commented on themes/default/content/registry/packages/kubernetes/how-to-guides/choosing-the-right-helm-resource-for-your-use-case.md:
```diff
@@ -0,0 +1,101 @@
+
+1. The ability to support transformations is limited to that offered by the `Helm` CLI (running `helm install` with `--post-renderer postrenderer`)
+2. Pulumi doesn't have direct access to the underlying Kubernetes resources installed by `Helm`, as a result, `CrossGuard` policies can't be enforced on Helm Release resources.
+3. Pulumi doesn't render fine-grained diffs in previews scoped to affected Kubernetes resources when Helm configuration changes.
```
```suggestion
3. Fine-grained preview support is limited, since the Helm client is responsible for managing the underlying Kubernetes resources.
```

lblackstone commented on themes/default/content/registry/packages/kubernetes/how-to-guides/choosing-the-right-helm-resource-for-your-use-case.md:
```diff
@@ -0,0 +1,101 @@
+
+#### *Fire-and-forget* Helm Chart Installation
+
+In many cases, users simply want to install an unmodified Chart and manage its configuration in their IaC tool of choice by specifying the chart and the values in code. While both the Chart and Release resources accomplish this with aplomb, we would recommend the use of Helm Release in such situations due to the broader support of Helm features such as Hooks.
```
```suggestion
In many cases, users simply want to install an unmodified Chart and manage its configuration in their IaC tool of choice by specifying the chart and the values in code. In this case, we recommend using Helm Release due to the broader support of Helm features such as hooks.
```

lblackstone commented on themes/default/content/registry/packages/kubernetes/how-to-guides/choosing-the-right-helm-resource-for-your-use-case.md:
```diff
@@ -0,0 +1,101 @@
+
+#### Interoperability with existing Helm releases
+
+If you have existing Helm Releases deployed through a version of the Helm CLI and wish to now integrate them in Pulumi, the `Helm Release` resource is your best choice. Currently component resources like `Chart` don't offer the ability to `import` existing resources.
```
```suggestion
If you have existing Helm Releases deployed through a version of the Helm CLI and wish to now integrate them in Pulumi, the `Helm Release` resource is your best choice. Currently, ComponentResources like `Chart` don't offer the ability to `import` existing resources.
```

lblackstone commented on themes/default/content/registry/packages/kubernetes/how-to-guides/choosing-the-right-helm-resource-for-your-use-case.md:
```diff
@@ -0,0 +1,101 @@
+
+`Chart` resources have direct access to the Kubernetes resources installed by the chart before installation. As a result, `Chart` resources support `transformations` which allow program authors to programmatically manipulate resources before they are installed by Pulumi. This is a very powerful tool which has enabled several advanced use cases for our users. Unfortunately, `Helm Release` does not ...
+
+Similarly, `Chart` resources can enumerate underlying resources and their inputs, thus providing fine-grained diffs on individual resources and richer previews.
```
```suggestion
Similarly, `Chart` resources can enumerate underlying resources and their inputs, thus providing fine-grained diffs and richer previews.
```

lblackstone commented on themes/default/content/registry/packages/kubernetes/how-to-guides/choosing-the-right-helm-resource-for-your-use-case.md:
```diff
@@ -0,0 +1,101 @@
+
+Similarly, `Chart` resources can enumerate underlying resources and their inputs, thus providing fine-grained diffs on individual resources and richer previews.
+
+If these are important to your environment or use case, then the `Helm Chart` resource should be preferred.
```
```suggestion
If these are important for your use case, then the `Helm Chart` resource is preferred.
```

lblackstone commented on themes/default/content/registry/packages/kubernetes/how-to-guides/choosing-the-right-helm-resource-for-your-use-case.md:
```diff
@@ -0,0 +1,101 @@
+
+#### Enforcing CrossGuard Policies on Kubernetes Resources
+
+`Chart` resources extract all Kubernetes objects and deploy them as Pulumi resources, making each of them privy to applicable CrossGuard policies. While resources installed by `Helm Release` are directly installed by Helm bypassing CrossGuard policies.
```
```suggestion
`Chart` resources extract all Kubernetes objects and deploy them as Pulumi resources, allowing fine-grained policy enforcement on these resources with CrossGuard. Since Pulumi does not manage the underlying resources from Helm Release, you should choose `Chart` if you need to enforce policy on these resources.
```

lblackstone commented on themes/default/content/registry/packages/kubernetes/how-to-guides/choosing-the-right-helm-resource-for-your-use-case.md:
```diff
@@ -0,0 +1,101 @@
+
+`Chart` resources extract all Kubernetes objects and deploy them as Pulumi resources, making each of them privy to applicable CrossGuard policies. While resources installed by `Helm Release` are directly installed by Helm bypassing CrossGuard policies.
+
+## Next Steps
+
+Now that you know which Helm resource is right for you, checkout some of the following resources for learning how to use each of them:
```
```suggestion
## Further reading

```

lblackstone reviewed (changes requested):
Overall content looks good.

viveklak commented:
@lblackstone I think I have addressed all your comments except a couple of cross links which I will add in another pass - thanks a bunch for your detailed review! Could you take another look through now to make sure the glaring issues are addressed?

viveklak commented on themes/default/content/blog/helm-release-resource-for-kubernetes-generally-available/index.md:
```diff
@@ -0,0 +1,415 @@
+---
+title: "Helm Release Resource for Kubernetes Generally Available"
+
+h1: "Announcing General Availability of Helm Release Resource for Kubernetes"
```
Fixed.

lblackstone commented on themes/default/content/registry/packages/kubernetes/how-to-guides/choosing-the-right-helm-resource-for-your-use-case.md:
```diff
@@ -0,0 +1,99 @@
+
+## Helm Chart Resources
+
+The [`Helm V2 Chart`]({{< relref "/registry/packages/kubernetes/api-docs/helm/v2/chart">}}), and [`Helm V3 Chart`]({{< relref "/registry/packages/kubernetes/api-docs/helm/v3/chart">}}) resources render the templates from your chart and then manage them directly with the Pulumi Kubernetes provider. These resources are implemented as [`Component Resources`]({{< relref "/docs/intro/concepts/resource ...
```
I'd suggest dropping the v2 mention here and in the intro. These links will go stale when the resource is removed.

viveklak replied:
Done.

stack72 reviewed (approved):
Ship it :)

viveklak commented:
This PR is now out of draft and can be considered finalized. The blog date has been set for Feb 4th (Friday) 0900 PST.
